runtime/doc/lua.txt

```diff
@@ -4433,4 +4433,250 @@ tohtml.tohtml({winid}, {opt})                         *tohtml.tohtml.tohtml()*
+                        `[userinfo "@"] host [":" port]`
+      • {credentials}?  (`string`) Proxy credentials with the format
+                        `username:password`
+
+*vim.net.curl.Metadata*
+    See the `--write-out` section in `man curl`
+
+    Fields: ~
+      • {certs}?                    (`string`) (Added in 7.88.0)
+      • {conn_id}?                  (`number`) (Added in 8.2.0)
+      • {content_type}?             (`string`)
+      • {errormsg}?                 (`string`) (Added in 7.75.0)
+      • {exitcode}?                 (`number`) (Added in 7.75.0)
+      • {filename_effective}?       (`string`) (Added in 7.26.0)
+      • {ftp_entry_path}?           (`string`)
+      • {headers}?                  (`table<string, string[]>`) (Added in
+                                    7.83.0) parsed result of ${header_json}
+      • {http_code}?                (`number`)
+      • {http_connect}?             (`number`)
+      • {http_version}?             (`string`) (Added in 7.50.0)
+      • {local_ip}?                 (`string`)
+      • {local_port}?               (`number`)
+      • {method}?                   (`string`) (Added in 7.72.0)
+      • {num_certs}?                (`number`) (Added in 7.88.0)
+      • {num_connects}?             (`number`)
+      • {num_headers}?              (`number`) (Added in 7.73.0)
+      • {num_redirects}?            (`number`)
+      • {num_retries}?              (`number`) (Added in 8.9.0)
+      • {proxy_ssl_verify_result}?  (`number`) (Added in 7.52.0)
+      • {proxy_used}?               (`number`) (Added in 8.7.0)
+      • {redirect_url}?             (`string`)
+      • {referer}?                  (`string`) (Added in 7.76.0)
+      • {remote_ip}?                (`string`)
+      • {remote_port}?              (`number`)
+      • {response_code}?            (`number`)
+      • {scheme}?                   (`string`) (Added in 7.52.0)
+      • {size_download}?            (`number`)
+      • {size_header}?              (`number`)
+      • {size_request}?             (`number`)
+      • {size_upload}?              (`number`)
+      • {speed_download}?           (`number`)
+      • {speed_upload}?             (`number`)
+      • {ssl_verify_result}?        (`number`)
+      • {time_appconnect}?          (`number`)
+      • {time_connect}?             (`number`)
+      • {time_namelookup}?          (`number`)
+      • {time_pretransfer}?         (`number`)
+      • {time_redirect}?            (`number`)
+      • {time_starttransfer}?       (`number`)
+      • {time_total}?               (`number`)
+      • {url}?                      (`string`) (Added in 7.75.0)
+      • {url.scheme}?               (`string`) (Added in 8.1.0)
+      • {url.user}?                 (`string`) (Added in 8.1.0)
+      • {url.password}?             (`string`) (Added in 8.1.0)
+      • {url.options}?              (`string`) (Added in 8.1.0)
+      • {url.host}?                 (`string`) (Added in 8.1.0)
```

Comment:
This looks like way too much. I mentioned in https://github.com/neovim/neovim/pull/29104#issuecomment-2144055981 and https://github.com/neovim/neovim/pull/23461#issuecomment-1660446261 that the first step should be a very simple interface. Adding all of these complex structures is a sign that something is going wrong.

The purpose is to download files. A simple use-case is downloading spellfiles, to start. There should be no need for elaborate structures.

Supporting all of curl's options is not maintainable. 

And proxy support is not something that is needed in the first PR. Please, just start with the common case.